text/0002-serializable-types.md

```diff
@@ -0,0 +1,413 @@
+
+We want to offer built-in serialization support for most commonly used types and also support custom serialization for user-defined types with ease.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+## Serialization
+
+Multiplayer framework has built-in serialization support for C# and Unity primitive types out-of-the-box, also with ability to further extend network serialization for user-defined types implementing `INetworkSerializable` interface.
+
+### C# Primitives
+
+`bool`, `char`, `sbyte`, `byte`, `short`, `ushort`, `int`, `uint`, `long`, `ulong`, `float`, `double`, `string` types will be serialized by built-in serialization code.
+
+```cs
+[ServerRpc]
+void FooServerRpc(int somenumber, string sometext) { /* ... */ }
+
+void Update()
+{
+    if (Input.GetKeyDown(KeyCode.P))
+    {
+        FooServerRpc(Time.frameCount, "hello, world"); // Client -> Server
+    }
+}
+```
+
+### Unity Primitives
+
+`Color`, `Color32`, `Vector2`, `Vector3`, `Vector4`, `Quaternion`, `Ray`, `Ray2D` types will be serialized by built-in serialization code.
+
+```cs
+[ClientRpc]
+void BarClientRpc(Color somecolor) { /* ... */ }
+
+void Update()
+{
+    if (Input.GetKeyDown(KeyCode.P))
+    {
+        BarClientRpc(Color.red); // Server -> Client
+    }
+}
+```
+
+### Enum Types
+
+A user-defined enum type will be serialized by built-in serialization code (with underlying integer type).
+
+```cs
+enum SmallEnum : byte
+{
+    A,
+    B,
+    C
+}
+
+enum NormalEnum // default -> int
+{
+    X,
+    Y,
+    Z
+}
+
+[ServerRpc]
+void ConfigServerRpc(SmallEnum smallEnum, NormalEnum normalEnum) { /* ... */ }
+
+void Update()
+{
+    if (Input.GetKeyDown(KeyCode.P))
+    {
+        ConfigServerRpc(SmallEnum.A, NormalEnum.X); // Client -> Server
+    }
+}
+```
+
+### Static Arrays
+
+Static arrays like `int[]` will be serialized by built-in serialization code if their underlying type is either one of serialization supported types (e.g. `Vector3`) or if they implement `INetworkSerializable` interface.
+
+```cs
+[ServerRpc]
+void HelloServerRpc(int[] scores, Color[] colors) { /* ... */ }
+
+[ClientRpc]
+void WorldClientRpc(MyComplexType[] values) { /* ... */ }
+```
+
+### NetworkObject & NetworkBehaviour
+
+`NetworkObject` and `NetworkBehaviour` instances will be serialized by built-in serialization code if instances are not `null` and `NetworkObject.IsSpawned == true`. Passing a `null` instance or a `NetworkObject` with `IsSpawned == false` or a `NetworkBehaviour` with `NetworkObject.IsSpawned == false` will still lead into RPC replication but remote side will have `null` values for the parameters. IDs of spawned `NetworkObject` and `NetworkBehaviour` instances will be resolved by running `NetworkManager` instance, and those IDs will be the links between local and remote instances. Also, those IDs will be used when serializing `NetworkObject` and `NetworkBehaviour` instances as a part of an RPC call.
```

Comment:
So I'm assuming it's not so much that the NetworkObject / NetworkBehaviour types are serialized (e.g. their contents are packed / unpacked in the RPC like the previous types).  Rather, you are transmitting to the recipient that this parameter is a NO / NB, and here's its ID that gets looked up by the NetworkManager and de-referenced on the receiving end.   That is, the actual transmission of the NO / NB state is happening through the NetworkingManager's Netvar mechanism, and this is just a way to pass a reference to a NO / NB.

If I have this right, I might suggest putting the mechanism like I described it first as when I first read this I thought "how is the serialization system magically serializing arbitrary types in my NO / NB without an INetworkSerializable interface.

Reply:
Yes, it's just ID-linking on the remote side and dereferences back to an already spawned NetworkObject and/or NetworkBehaviour. I will rework this section with better wording and outline the actual mechanism as you suggested!